pkg/broker/ingress/multi_topic_decouple_sink_test.go

```diff
@@ -0,0 +1,210 @@
+				},
+			},
+			cases: []brokerTestCase{
+				{
+					ns:          "test_ns_1",
+					broker:      "test_broker_1",
+					topic:       "test_topic_1",
+					clientErrFn: injectErr("test_topic_1", errors.New("inject error")),
+					wantErr:     true,
+				},
+			},
+		},
+		{
+			name:         "broker doesn't exist in config",
+			brokerConfig: &config.TargetsConfig{},
+			cases: []brokerTestCase{
+				{
+					ns:      "test_ns_1",
+					broker:  "test_broker_1",
+					topic:   "test_topic_1",
+					wantErr: true,
+				},
+			},
+		},
+		{
+			name: "decouple queue is nil for broker",
+			brokerConfig: &config.TargetsConfig{
+				Brokers: map[string]*config.Broker{
+					"test_ns_1/test_broker_1": {DecoupleQueue: nil,},
+				},
+			},
+			cases: []brokerTestCase{
+				{
+					ns:      "test_ns_1",
+					broker:  "test_broker_1",
+					topic:   "test_topic_1",
+					wantErr: true,
+				},
+			},
+		},
+		{
+			name: "empty topic for broker",
+			brokerConfig: &config.TargetsConfig{
+				Brokers: map[string]*config.Broker{
+					"test_ns_1/test_broker_1": {DecoupleQueue: &config.Queue{Topic: "",},},
+				},
+			},
+			cases: []brokerTestCase{
+				{
+					ns:      "test_ns_1",
+					broker:  "test_broker_1",
+					topic:   "test_topic_1",
+					wantErr: true,
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fakeClient := newFakePubsubClient(t)
+			brokerConfig := memory.NewTargets(tt.brokerConfig)
+			for _, testCase := range tt.cases {
+				if testCase.clientErrFn != nil {
+					testCase.clientErrFn(fakeClient)
+				}
+				sink, err := NewMultiTopicDecoupleSink(context.Background(), WithBrokerConfig(brokerConfig), WithPubsubClient(fakeClient))
+				if err != nil {
+					t.Fatalf("Failed to create decouple sink: %v", err)
+				}
+
+				// Send events
+				event := createTestEvent(uuid.New().String())
+				err = sink.Send(context.Background(), testCase.ns, testCase.broker, *event)
+
+				// Verify results.
+				if testCase.wantErr && err == nil {
+					t.Fatal("Want error but got nil")
+				}
+				if !testCase.wantErr && err != nil {
+					t.Fatalf("Unexpected error: %v", err)
+				}
+				if !testCase.wantErr {
+					got := <-fakeClient.topics[testCase.topic]
+					if dif := cmp.Diff(*event, got); dif != "" {
+						t.Errorf("Output event doesn't match input, dif: %v", dif)
+					}
+				}
+			}
+		})
+	}
+}
+
+type fakePubsubClient struct {
+	t *testing.T
+	// topics is the mapping from topic name to corresponding channel which contains the event.
+	topics        map[string]<-chan cloudevents.Event
+	topicToClient map[string]cloudevents.Client
+	// topicToErr injects error to a topic to simulate client error.
+	topicToErr map[string]error
+}
+
+func newFakePubsubClient(t *testing.T) *fakePubsubClient {
+	return &fakePubsubClient{
+		t:             t,
+		topics:        map[string]<-chan cloudevents.Event{},
+		topicToClient: map[string]cloudevents.Client{},
+		topicToErr:    map[string]error{},
+	}
+}
```

Comment:
Would it be possible to use pubsub test client `pstest` package. I think it would be more close to real case.

Reply:
I gave it a try and it turned out there is a lot of code required to set up the client, topics, subscriptions, and retrieve events from topics. Essentially we have to do many jobs that the controller does, plus plumbing the test client to cloudevents client.

I agree it's more "real" to use that but meanwhile the added complexity makes the test harder to read. Moreover, the decouple sink relies on a cloudevent client abstraction, not necessarily pubsub. So I prefer to keep the test as is.

Reply:
It's really not that much code. I have a one written in my fanout service PR. It's like 15 lines more? Due to the fact we lack test coverage in the cloudevents sdk for pubsub, I still highly recommend using pstest :)

Reply:
There is also an issue with cloudevents v2 client. If you cancel the context, it panics. There is a TODO to handle error properly in ce v2. Due to that, it gets ugly to receive events back from pubsub. I guess there are workarounds to use the binding instead of the client, but then there is more logic not related to the test itself.

That said, I agree we need some more real testing. In fact, it's on my list to write some integration test using real pubsub (or pstest, which can be swapped easily). The test will be more like a blackbox test for the ingress as a whole. Does that sound good to you?

Reply:
Created an issue to follow up on this: https://github.com/google/knative-gcp/issues/786

Reply:
Yes. Just a note. If we're going to use pstest, it's better to replace this one here.